modules/storages/lib/api/v3/file_links/create_endpoint.rb

```diff
@@ -0,0 +1,84 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2022 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module API::V3::FileLinks
+  # Handles /api/v3/work_packages/:work_package_id/file_links as defined
+  # in modules/storages/lib/api/v3/file_links/work_packages_file_links_api.rb
+  #
+  # Multiple classes are involved during its lifecycle:
+  #   - API::V3::FileLinks::ParseCreateParamsService
+  #   - API::V3::FileLinks::FileLinkCollectionRepresenter
+  #   - Storages::FileLinks::CreateService
+  #
+  # These classes are either deduced from the model class, or given as parameter
+  # on class instantiation.
+  class CreateEndpoint < API::Utilities::Endpoints::Create
+    include ::API::V3::Utilities::Endpoints::V3Deductions
+    include ::API::V3::Utilities::Endpoints::V3PresentSingle
+
+    # As this endpoint receives a list of file links to create, it calls the
+    # create service multiple times, one time for each file link to create. The
+    # call is done by calling the `super` method. Results are aggregated in
+    # global_result using the `add_dependent!` method.
+    def process(request, params_elements)
+      global_result = ServiceResult.new(success: true)
+
+      Storages::FileLink.transaction do
+        params_elements.each do |params|
+          # call the default API::Utilities::Endpoints::Create#process
+          # implementation for each of the params_element array
+          one_result = super(request, params)
+          # merge service result in one
+          global_result.add_dependent!(one_result)
+        end
+
+        # rollback records created if an error occurred (validation failed)
+        raise ActiveRecord::Rollback if global_result.failure?
+      end
```

Comment:
This should work but it made me wonder whether we should add `requires: new` to the transaction blocks we have defined within our services to ensure that a service is atomic and does not lead to unwanted commits in case they are wrapped inside a transaction (see https://www.bigbinary.com/learn-rubyonrails-book/activerecord-transactions-in-depth#nested-transactions)

Here however, with the transaction block raising its own `ActiveRecord::Rollback`, it should be safe.

Reply:
Indeed, it looks better to create a transaction, especially when a service calls another service. I'll be happy to add a `requires_new: true` if you can write a unit test that demonstrates your point.

Also I do not think it's particularly related to this pull request, as the issue can happen with services calling other services in their `perform`.

Reply:
Right, that is definitely not releated to the PR, so it does not have to be adapted in it. I was just made aware of it yesterday when I basically did:

```
Model.transaction do 
  Model.transaction do
    [Change database]
    
    raise ActiveRecord::Rollback
  end
end
```
and the change I anticipated gone was still there. 

I'll try to make the necessary changes in a separate PR.